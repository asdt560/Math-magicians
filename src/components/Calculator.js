import React, { useState } from 'react';
import calculate from '../logic/calculate';

const Calculator = () => {
  const [state, setState] = useState({
    total: '0',
    next: '0',
    operation: null,
  });
  const onClickHandler = (e) => {
    if (e.target.innerHTML.match(/[0-9]+/)) {
      setState((prevState) => ({
        total: prevState.next,
        next: e.target.innerHTML,
      }));
    } else {
      setState({ operation: e.target.innerHTML });
    }
    const result = calculate(state, e.target.innerHTML);
    setState(result);
  };
  let result = '0';
  if (state.total !== null) {
    result = state.total;
  } else if (state.total === null && state.next === null) {
    result = '0';
  } else {
    result = state.next;
  }
  return (
    <section className="calculator">
      <div className="row top">
        { result }
      </div>
      <div className="row">
        <button className="button" type="button" onClick={onClickHandler}>AC</button>
        <button className="button" type="button" onClick={onClickHandler}>+/-</button>
        <button className="button" type="button" onClick={onClickHandler}>%</button>
        <button className="button orange" type="button" onClick={onClickHandler}>รท</button>
      </div>
      <div className="row">
        <button className="button" type="button" onClick={onClickHandler}>7</button>
        <button className="button" type="button" onClick={onClickHandler}>8</button>
        <button className="button" type="button" onClick={onClickHandler}>9</button>
        <button className="button orange" type="button" onClick={onClickHandler}>x</button>
      </div>
      <div className="row">
        <button className="button" type="button" onClick={onClickHandler}>4</button>
        <button className="button" type="button" onClick={onClickHandler}>5</button>
        <button className="button" type="button" onClick={onClickHandler}>6</button>
        <button className="button orange" type="button" onClick={onClickHandler}>-</button>
      </div>
      <div className="row">
        <button className="button" type="button" onClick={onClickHandler}>1</button>
        <button className="button" type="button" onClick={onClickHandler}>2</button>
        <button className="button" type="button" onClick={onClickHandler}>3</button>
        <button className="button orange" type="button" onClick={onClickHandler}>+</button>
      </div>
      <div className="row">
        <button className="button zero" type="button" onClick={onClickHandler}>0</button>
        <button className="button" type="button" onClick={onClickHandler}>.</button>
        <button className="button orange" type="button" onClick={onClickHandler}>=</button>
      </div>
    </section>
  );
};

export default Calculator;
