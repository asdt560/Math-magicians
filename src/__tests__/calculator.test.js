import '@testing-library/jest-dom';
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Calculator from '../components/Calculator';

describe('calculator.js tests', () => {
  it('calculator render test', () => {
    render(<Calculator />);
    const result = document.querySelector('.calculator');
    expect(result).toMatchSnapshot();
  });

  it('calculator interaction test - add a number', async () => {
    render(<Calculator />);
    await userEvent.click(screen.getByText('9'));
    const result = document.querySelector('.top');
    expect(result).toHaveTextContent('9');
  });

  it('calculator interaction test - perform operation', async () => {
    render(<Calculator />);
    const nine = screen.getByText('9');
    await userEvent.click(nine);
    await userEvent.click(screen.getByText('x'));
    await userEvent.click(nine);
    await userEvent.click(screen.getByText('='));
    const result = document.querySelector('.top');
    expect(result).toHaveTextContent('81');
  });
});
