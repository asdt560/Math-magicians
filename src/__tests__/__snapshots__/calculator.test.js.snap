// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator.js tests calculator render test 1`] = `
<section
  class="calculator"
>
  <div
    class="row top"
  >
    0
  </div>
  <div
    class="row"
  >
    <button
      class="button"
      type="button"
    >
      AC
    </button>
    <button
      class="button"
      type="button"
    >
      +/-
    </button>
    <button
      class="button"
      type="button"
    >
      %
    </button>
    <button
      class="button orange"
      type="button"
    >
      รท
    </button>
  </div>
  <div
    class="row"
  >
    <button
      class="button"
      type="button"
    >
      7
    </button>
    <button
      class="button"
      type="button"
    >
      8
    </button>
    <button
      class="button"
      type="button"
    >
      9
    </button>
    <button
      class="button orange"
      type="button"
    >
      x
    </button>
  </div>
  <div
    class="row"
  >
    <button
      class="button"
      type="button"
    >
      4
    </button>
    <button
      class="button"
      type="button"
    >
      5
    </button>
    <button
      class="button"
      type="button"
    >
      6
    </button>
    <button
      class="button orange"
      type="button"
    >
      -
    </button>
  </div>
  <div
    class="row"
  >
    <button
      class="button"
      type="button"
    >
      1
    </button>
    <button
      class="button"
      type="button"
    >
      2
    </button>
    <button
      class="button"
      type="button"
    >
      3
    </button>
    <button
      class="button orange"
      type="button"
    >
      +
    </button>
  </div>
  <div
    class="row"
  >
    <button
      class="button zero"
      type="button"
    >
      0
    </button>
    <button
      class="button"
      type="button"
    >
      .
    </button>
    <button
      class="button orange"
      type="button"
    >
      =
    </button>
  </div>
</section>
`;
